Kit with preferredGenerator isn't finding generator

Reproduce issue like this:
Install jom (https://wiki.qt.io/Jom).
Add jom.exe location to your system PATH.
In cmake-tools-kits.json add the following kit:
`  {"name": "JOM test kit", "preferredGenerator": {"name": "NMake Makefiles JOM"}}`

Open a CMake project and try to configure it with the "JOM test kit".
You will get the error:

> [main] Unable to determine what CMake generator to use. Please install or configure a preferred generator, or update settings.json, your Kit configuration or PATH variable. Error: No usable generator found.

_Originally posted by @KorkiFork in https://github.com/microsoft/vscode-cmake-tools/issues/4019#issuecomment-2318895439_

Still not working.

Diagnostic log:

```json
{
  "os": "win32",
  "vscodeVersion": "1.92.2",
  "cmtVersion": "1.13.0",
  "configurations": [
    {
      "folder": "c:\\Users\\marco\\Documents\\dev\\random_folder",
      "cmakeVersion": "3.30.0",
      "configured": true,
      "generator": "Ninja",
      "usesPresets": true,
      "compilers": {
        "C": "C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.36.32532/bin/Hostx64/x64/cl.exe",
        "CXX": "C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.36.32532/bin/Hostx64/x64/cl.exe"
      }
    }
  ],
  "cpptoolsIntegration": {
    "isReady": false,
    "hasCodeModel": false,
    "activeBuildType": "",
    "buildTypesSeen": [],
    "requests": [],
    "responses": [],
    "partialMatches": [],
    "targetCount": 0,
    "executablesCount": 0,
    "librariesCount": 0,
    "targets": []
  },
  "settings": [
    {
      "communicationMode": "automatic",
      "useCMakePresets": "auto",
      "configureOnOpen": false
    }
  ]
}
```

Build log:
```log
vcpkg_installed -DVCPKG_OVERLAY_TRIPLETS=C:/Users/marco/Documents/dev/random_folder/3rd_party/triplets -DCMAKE_TOOLCHAIN_FILE=C:/Users/marco/Documents/dev/games/git/vcpkg/scripts/buildsystems/vcpkg.cmake -DCMAKE_INSTALL_PREFIX=C:/Users/marco/Documents/dev/random_folder/out/install/windows-debug -SC:/Users/marco/Documents/dev/random_folder -BC:/Users/marco/Documents/dev/random_folder/out/build/windows-debug -G Ninja
[cmake] -- Running vcpkg install
[cmake] All requested packages are currently installed.
[cmake] Total install time: 200 ns
[cmake] bext-di is header-only and can be used from CMake via:
...
...
[cmake]   find_package(Qt6Xml CONFIG REQUIRED)
[cmake]   target_link_libraries(main PRIVATE Qt::Xml Qt6::Xml Qt::XmlPrivate Qt6::XmlPrivate)
...
...
[cmake] -- Running vcpkg install - done
[cmake] -- The C compiler identification is unknown
[cmake] -- The CXX compiler identification is unknown
[cmake] CMake Error at CMakeLists.txt:21 (project):
[cmake]   The CMAKE_C_COMPILER:
[cmake] 
[cmake]     cl.exe
[cmake] 
[cmake]   is not a full path and was not found in the PATH.  Perhaps the extension is
[cmake]   missing?
[cmake] 
[cmake]   Tell CMake where to find the compiler by setting either the environment
[cmake]   variable "CC" or the CMake cache entry CMAKE_C_COMPILER to the full path to
[cmake]   the compiler, or to the compiler name if it is in the PATH.
[cmake] 
[cmake] 
[cmake] CMake Error at CMakeLists.txt:21 (project):
[cmake]   The CMAKE_CXX_COMPILER:
[cmake] 
[cmake]     cl.exe
[cmake] 
[cmake]   is not a full path and was not found in the PATH.  Perhaps the extension is
[cmake]   missing?
[cmake] 
[cmake]   Tell CMake where to find the compiler by setting either the environment
[cmake]   variable "CXX" or the CMake cache entry CMAKE_CXX_COMPILER to the full path
[cmake]   to the compiler, or to the compiler name if it is in the PATH.
[cmake] 
[cmake] 
[cmake] -- Configuring incomplete, errors occurred!
[proc] The command: "C:\Program Files\CMake\bin\cmake.EXE" -DCMAKE_BUILD_TYPE=Debug -DCMAKE_C_COMPILER=cl.exe -DCMAKE_CXX_COMPILER=cl.exe -DX_VCPKG_APPLOCAL_DEPS_INSTALL=1 -DVCPKG_INSTALLED_DIR=C:/Users/marco/Documents/dev/random_folder/out/vcpkg_installed -DVCPKG_OVERLAY_TRIPLETS=C:/Users/marco/Documents/dev/random_folder/3rd_party/triplets -DCMAKE_TOOLCHAIN_FILE=C:/Users/marco/Documents/dev/games/git/vcpkg/scripts/buildsystems/vcpkg.cmake -DCMAKE_INSTALL_PREFIX=C:/Users/marco/Documents/dev/random_folder/out/install/windows-debug -SC:/Users/marco/Documents/dev/random_folder -BC:/Users/marco/Documents/dev/random_folder/out/build/windows-debug -G Ninja exited with code: 1
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -2,9 +2,10 @@
 
 ## 1.19.50
 
-Bug Fixes: 
+Bug Fixes:
 
 - Fix env expansion of all variables (toolchainFile, etc.) in presets. [#4019](https://github.com/microsoft/vscode-cmake-tools/issues/4019)
+- Fix generator and preferredGenerator logic. [#4031](https://github.com/microsoft/vscode-cmake-tools/issues/4031), [#4005](https://github.com/microsoft/vscode-cmake-tools/issues/4005), [#4032](https://github.com/microsoft/vscode-cmake-tools/issues/4032)
 
 ## 1.19.49
 

--- a/src/drivers/cmakeDriver.ts
+++ b/src/drivers/cmakeDriver.ts
@@ -744,17 +744,17 @@ export abstract class CMakeDriver implements vscode.Disposable {
             preferredGenerators.push({ name: "Unix Makefiles" });
         }
 
-        // If a generator is set in the "cmake.generator" setting, push it to the front
-        // of the "best generator" logic
+        // Use the "best generator" logic only if the user did not define a particular
+        // generator to be used via the `cmake.generator` setting.
         if (this.config.generator) {
-            preferredGenerators.unshift({
+            this._generator = {
                 name: this.config.generator,
                 platform: this.config.platform || undefined,
                 toolset: this.config.toolset || undefined
-            });
+            };
+        } else {
+            this._generator = await this.findBestGenerator(preferredGenerators);
         }
-
-        this._generator = await this.findBestGenerator(preferredGenerators);
     }
 
     protected abstract doSetConfigurePreset(needsClean: boolean, cb: () => Promise<void>): Promise<void>;
@@ -1004,7 +1004,7 @@ export abstract class CMakeDriver implements vscode.Disposable {
                 return false;
             })();
             if (!generator_present) {
-                const vsMatch = /^(Visual Studio \d{2} \d{4})($|\sWin64$|\sARM$)/.exec(gen.name);
+                const vsMatch = /^(Visual Studio \d{2} \d{4})($|\sWin64$|\sARM$)/.exec(gen_name);
                 if (platform === 'win32' && vsMatch) {
                     let possibleArchitecture = vsMatch[2].trim();
                     if (possibleArchitecture && possibleArchitecture === "Win64") {
@@ -1016,11 +1016,17 @@ export abstract class CMakeDriver implements vscode.Disposable {
                         toolset: gen.toolset
                     };
                 }
-                if (gen.name.toLowerCase().startsWith('xcode') && platform === 'darwin') {
+                if (gen_name.toLowerCase().startsWith('xcode') && platform === 'darwin') {
                     return gen;
                 }
 
-                // If the generator isn't found, move on to the next one
+                // If it is not a common generator that we can find, but it is a known cmake generator (cmakeGenerators), return it.
+                // The only caveat is that we should not return a Visual Studio generator on non-Windows platforms.
+                if (this.cmakeGenerators.indexOf(gen_name) >= 0 && !this.isCommonGenerator(gen.name)) {
+                    if (gen_name.startsWith("Visual Studio") && platform === "win32") {
+                        return gen;
+                    }
+                }
                 continue;
             } else {
                 return gen;

--- a/src/drivers/cmakeDriver.ts
+++ b/src/drivers/cmakeDriver.ts
@@ -1022,10 +1022,8 @@ export abstract class CMakeDriver implements vscode.Disposable {
 
                 // If it is not a common generator that we can find, but it is a known cmake generator (cmakeGenerators), return it.
                 // The only caveat is that we should not return a Visual Studio generator on non-Windows platforms.
-                if (this.cmakeGenerators.indexOf(gen_name) >= 0 && !this.isCommonGenerator(gen.name)) {
-                    if (!(gen_name.startsWith("Visual Studio") && platform === "win32")) {
-                        return gen;
-                    }
+                if (this.cmakeGenerators.indexOf(gen_name) >= 0 && !this.isCommonGenerator(gen.name) && !(gen_name.startsWith("Visual Studio") && platform !== "win32")) {
+                    return gen;
                 }
                 continue;
             } else {

--- a/src/drivers/cmakeDriver.ts
+++ b/src/drivers/cmakeDriver.ts
@@ -1023,7 +1023,7 @@ export abstract class CMakeDriver implements vscode.Disposable {
                 // If it is not a common generator that we can find, but it is a known cmake generator (cmakeGenerators), return it.
                 // The only caveat is that we should not return a Visual Studio generator on non-Windows platforms.
                 if (this.cmakeGenerators.indexOf(gen_name) >= 0 && !this.isCommonGenerator(gen.name)) {
-                    if (gen_name.startsWith("Visual Studio") && platform === "win32") {
+                    if (!(gen_name.startsWith("Visual Studio") && platform === "win32")) {
                         return gen;
                     }
                 }

--- a/src/drivers/cmakeDriver.ts
+++ b/src/drivers/cmakeDriver.ts
@@ -1022,7 +1022,7 @@ export abstract class CMakeDriver implements vscode.Disposable {
 
                 // If it is not a common generator that we can find, but it is a known cmake generator (cmakeGenerators), return it.
                 // The only caveat is that we should not return a Visual Studio generator on non-Windows platforms.
-                if (this.cmakeGenerators.indexOf(gen_name) >= 0 && !this.isCommonGenerator(gen.name) && !(gen_name.startsWith("Visual Studio") && platform !== "win32")) {
+                if (this.cmakeGenerators.indexOf(gen_name) >= 0 && !this.isCommonGenerator(gen_name) && !(gen_name.startsWith("Visual Studio") && platform !== "win32")) {
                     return gen;
                 }
                 continue;

